pyiron/atomistics/structure/visualization.py

```diff
@@ -0,0 +1,604 @@
+# coding: utf-8
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+
+import numpy as np
+import warnings
+from matplotlib.colors import rgb2hex
+from pyiron_base import Settings
+from scipy.interpolate import interp1d
+
+__author__ = "Joerg Neugebauer, Sudarsan Surendralal"
+__copyright__ = (
+    "Copyright 2020, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "1.0"
+__maintainer__ = "Sudarsan Surendralal"
+__email__ = "surendralal@mpie.de"
+__status__ = "production"
+__date__ = "Sep 1, 2017"
+
+s = Settings()
+
+
+def plot3d_plotly(
+    atoms,
+    scalar_field=None,
+    particle_size=1.0,
+    camera="orthographic",
+    view_plane=np.array([1, 1, 1]),
+    distance_from_camera=1.25,
+    opacity=1,
+):
+    """
+    Make a 3D plot of the atomic structure.
+
+    Args:
+        camera (str): 'perspective' or 'orthographic'. (Default is 'perspective'.)
```

Comment:
I think you are missing the `atoms` argument in the docstring. (Also found elsewhere)

Reply:
This is a bit intentional though, as the `atoms` argument is just `self` when the method gets bound to the `Atoms` class instance. I thought there was a nice middle ground that the plain english sentence at the head of the docstring makes the context of the `atoms` argument pretty clear. We maybe need something even better than this though... Perhaps something at the module level documentation of `visualization.py` giving a generic definition of the `atoms` argument and explaining the intent of binding these methods?

Reply:
I've been thinking about it all the time and I still don't know what's the best way. I'm right now slightly inclined to make the function private and not to put `atoms` in the DocString. At least in the current format, @sudarsan-surendralal is right that as long as the function is independently functional, there must be an entry for `atoms`, even though we know that it will be confusing in most of the cases since it will be mostly used from `Atoms`.

Reply:
I think making the functions private is bad style, since we use it elsewhere, but I am also in agreement with Sudarsan's fundamental complaint. What happens if we make the whole visualization module private? Do we get style complaints when a sibling module inherits it with a relative path import call? I am happy to check, but I will have to wait until little mister is asleep to focus that deeply.

Reply:
> What happens if we make the whole visualization module private?

What do you mean by that?

Reply:
> > What happens if we make the whole visualization module private?
> 
> What do you mean by that?

@samwaseda yes, it works as I hoped!

In `pyiron/atomistics/structure` I added a module `_foo.py` containing a single method `bar`.

Then in `atoms.py` I can write `from . _foo import bar` and pycharm is perfectly happy. But over in, say, a random file in pyiron_contrib, if I write `from pyiron.atomistics.structure._foo import bar` the IDE gives me the warning that I am using "Access to a protected member _foo of a module".

Basically, by flagging the entire sub-module with an underscore we're making it private from outside the scope of its parent module. It's python, so consenting adults can access it if they want, but then I feel comfortable that we're no longer responsible for the fact that the docstrings don't make sense outside the context they were intended for.

@sudarsan-surendralal how do you feel about this?

Reply:
I'm fine with the idea, but I don't really understand why it's bad to put `def _plot3d`.

Reply:
If you put 'def _plot3d' in one module, then import it in another module, any decent IDE will complain that you're accessing a private variable. The leading underscore should indicate that the function (/class/module/method/variable) is only intended to be used within the current scope (i.e. the same module/module/parent module/class/module for function/class/module/method/variable, respectively). 

It will break absolutely nothing to use `def _plot3d`, it's just not conventional. Well, I guess it might break the codacy check, but whatever.

If we're going to use an underscore though, I'd rather underscore the module, because it's better form.
cf. the top answer in [this SO question](https://stackoverflow.com/questions/3602110/python-private-module-in-a-package), which backs up what my PyCharm 'experiment' showed me.

Reply:
Actually, the answer from that SO question that directs to the [relevant passage of PEP8: public and internal interfaces](https://www.python.org/dev/peps/pep-0008/?#id50) is even better

Reply:
Ah ok I didn't know that.

Reply:
I agree with @liamhuber, it's better to avoid using private functions in this case. This is because private functions are meant to be used only within the module (in this case: `visualization.py`) and therefore it's not good practice to import these private functions elsewhere (eg. `atoms.py`). I like the private module option that Liam mentioned (`_foo.py`). I haven't used this before but this is exactly what we want, public functions that are accessible to developers to be used elsewhere in the codebase but not to the users. 

Regarding the docstrings, I think it's nice to have `atoms` described as an argument even though it might seem redundant just for the sake of consistency (and to stop IDEs like pycharm from complaining!).

Reply:
So for me the best options are:

- `visualize.py` and docstring entry for `atoms`
- `_visualize.py` and **no** docstring entry for `atoms`

Frankly I'm fine with both options but I'd be rather unwilling to do anything other than those.

Reply:
Agreed. Since Sudarsan is on board with the private module, and it's a nifty part of python we haven't used yet, let's go with that ;)

Reply:
I do think the module-level docstring in `_visualize.py` should have sentence to make clear the intention that those methods get bound to an Atoms object though.

Reply:
Wait, if there's a DocString entry for `atoms`, then there's no reason to make `visualize.py` private, right?

Reply:
Oh sorry so do I understand correctly that @sudarsan-surendralal wants a DocString entry for `atoms` and @liamhuber doesn't?

Reply:
I propose your second solution: `_visualize.py`, where the public functions therein don't have a docstring element for `atoms`. However, I would additionally like to add to the head of `_visualize.py` a module-level docstring like 

"""
This module encapsulates visualization techniques for the `Atoms` class. The public facing methods here are intended to be bound to that class, and thus the first argument of these is always of types `Atoms` and should not appear in the docstrings.
"""

This fits between the imports at the `__author__` declaration for consistency with other modules.

Reply:
Yes and that's where I see a difference between you two, since from what I understand @sudarsan-surendralal wants `atoms` explicitly in the DocString. I personally support your opinions equally, so you guys can decide what's better.

Reply:
I just committed what could be a compromise -> I put `plot3d` etc. in the class `Visualize(object)`, which does nothing, but the first argument is now `self`. The rest is essentially the same (i.e. instead of `MethodType(plot3d, self)` I put `MethodType(Visualize.plot3d, self)` etc.) What do you think?

Reply:
And since this class cannot be used standalone I indeed made a private module. Let me know what you think.

Reply:
This comes back a little bit to my complaint that a class without data probably shouldn't be a class.

What if we just rename the `atoms` argument as `self`? I think this makes the intention maximally clear.

Reply:
Maybe I'm barking up the wrong tree. Let's google around for a minute what common solutions are to too-long classes.

Reply:
I'd claim it is a class with data, it's just this class itself does not provide data.

I guess pycharm will anyway complain whether it's called `atoms` or `self`. And I agree with @sudarsan-surendralal that as long as it's a function, all arguments should be listed in the DocString.

Reply:
@samwaseda Sorry for causing so much trouble when you're just trying to clean things up!!! I think this sort of cleanup is needed in many places right now, so my motivation was just to give you a really hard time this once so we get a good transferable solution worked out ;)

Reply:
> I'd claim it is a class with data, it's just this class itself does not provide data.
Agreed. I just thought with only a single piece of data it might be better as a module rather than a class. I definitely agree that we wind up running into docstring issues by going down that path though.


I couldn't find anything on the internet, but in Fowler's 'Refactoring', breaking off a new class to handle part of the class that originally had too much responsibility is indeed the suggested solution.

In the Fowler-recommended solution there's no rebinding of `object.sub_object.method` to `object.method`, but I guess for us `structure.plot3d` *must* be rebound for backwards compatibility.

So, frustratingly, at the end of the day I come back to most favouring your very original formulation. I still am not 100% happy with it, but it's still very clearly forward progress -- I think all the solutions we've discussed are, but the original one is probably the best/safest...

Reply:
> @samwaseda Sorry for causing so much trouble when you're just trying to clean things up!!! I think this sort of cleanup is needed in many places right now, so my motivation was just to give you a really hard time this once so we get a good transferable solution worked out ;)

No worries - this one is an important conceptual question we should discuss to create a common ground. I find it particularly crucial in the absence of personal discussions, because now we are vulnerable to the danger that we keep doing things according to our own individual tastes, which will certainly make pyiron fall apart at some point. If we disagree on certain issues, we should definitely discuss, and I'd like to strongly encourage everyone to follow your example.

Reply:
> So, frustratingly, at the end of the day I come back to most favouring your very original formulation. I still am not 100% happy with it, but it's still very clearly forward progress -- I think all the solutions we've discussed are, but the original one is probably the best/safest...

The original formulation being the current version, correct?

Reply:
One thing I could do is to add

```
def __init__(self):
    raise ValueError('This class cannot be used standalone')
```

With this, `Visualize.plot3d(structure)` will still work, but it won't be possible to do things like `view = Visualize()`.

Reply:
> No worries - ...

Cool, yes I entirely agree! I just wanted to be super clear that I am grateful for the effort you're putting in and am not simply trying to be a pain in the ass ;) I also think you're spot on on making sure that we have (at least a moderately) shared vision for how these things should look!

Reply:
> The original formulation being the current version, correct?

No, rather

```python
# _visualization.py
class Visualize:
    def __init__(self, atoms):
        self._atoms = atoms
    
#atoms.py 
class Atoms(ASEAtoms):
    def __init__(self):
        self.visualize = Visualize(self)
```

Reply:
Really? This will create 100 duplicate lines. In terms of reducing the size of `atoms.py`, that's quite inefficient, especially if it only boils down to the DocString problem. And from what I understand, Fowler is talking about class functions, which **must** retain a class structure. But in this case, as you correctly pointed out, the functionalities are largely independent, so the solution we have here does not go against what he describes (I'll gladly accept criticism though :D)

Reply:
Gah! No! I don't want 100 duplicate lines ;)
I must not have communicated clearly. I mean that I now have come full circle to embrace your original version where all the plotting methods are pulled out of `Atoms` and put into `visualization.Visualization`. However, unlike the current setup where the type of `Visualization` needs to implicitly be `Atoms`, I'm proposing that we go back to the old setup where it's just a regular class with a data field of type `Atoms`, then inside `Atoms` we instantiate it (`self.visuazation = Visualization(self)`, and wrap it's plot functions `def plot3d(self, *args, **kwargs): self.visualization.plot3d(*args, **kwargs)`.

I guess then we have a handful of dull lines where we wrap the plotting functions, but since we are only making quick aliases for the few public functions of `Visualize` it should be pretty short.

In fact, reading the further conversation it might be nice if `Visualize` had separate methods like `_plot3d_ase` and `_plot3d_plotly` (or whatever!), then a *single* public method `plot3d` which also takes the `method` kwarg. Then we'd only need to alias a *single* method in `Atoms`, and we could change the default plotting behaviour to use `plotly` instead of NGLView without borking the demo notebooks.

Reply:
ah ok it's a good idea with `*args`. I'll do it this way.

I guess for `plotly` to be the default there are still quite some changes to make (`vector_field`, plotting frame etc.), so I'll leave this point for a later PR.

Reply:
I guess now it's done.